### PR TITLE
Loose coupling nsgaii elite population selection

### PR DESCRIPTION
**Pull request by Alnusjaponica**
## Motivation
Progress v3.3 task to make `NSGAII Sampler` loose coupled.

## Description of the changes
Add `_elite_population_selection_strategy` argument to the constructor if `NSGAIISampler` and separate the elite population selection process so that users can customize surviving process.

Please merge this PR after https://github.com/optuna/optuna/pull/4806 is merged.

### CODE REVIEW

**Comment by toshihikoyanase** [comment]
@HideakiImamura Could you review this PR, please?

File: optuna/samplers/_nsgaiii.py
Comment (HideakiImamura) [inline optuna/samplers/_nsgaiii.py:1]: Just for confirmation. The loose coupling of the elite population selection for the `NSGAIIISampler` will be done as a follow-up?

File: optuna/samplers/nsgaii/_sampler.py [hunk 1]
```diff
@@ -119,6 +118,10 @@ def __init__(
         swapping_prob: float = 0.5,
         seed: int | None = None,
         constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
+        elite_population_selection_strategy: Callable[
```
Comment (HideakiImamura) [inline optuna/samplers/nsgaii/_sampler.py:121]: Could you add a docstring for this argument?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have two comments. PTAL.

File: optuna/samplers/_nsgaiii.py
Comment (Alnusjaponica) [reply optuna/samplers/_nsgaiii.py:1]: Yes, I am going to do it as a follow-up.

File: optuna/samplers/nsgaii/_sampler.py (hunk 1 above)
Comment (Alnusjaponica) [reply optuna/samplers/nsgaii/_sampler.py:129]: Thanks for the comment. I added the explanation.

File: optuna/samplers/_nsgaiii.py [hunk 2]
```diff
@@ -102,17 +102,6 @@ def __init__(
...
-            raise ValueError(
-                "`mutation_prob` must be None or a float value within the range [0.0, 1.0]."
-            )
-
-        if not (0.0 <= crossover_prob <= 1.0):
-            raise ValueError("`crossover_prob` must be a float value within the range [0.0, 1.0].")
-
-        if not (0.0 <= swapping_prob <= 1.0):
-            raise ValueError("`swapping_prob` must be a float value within the range [0.0, 1.0].")
-
```
Comment (toshihikoyanase) [inline optuna/samplers/_nsgaiii.py:115]: [Question] I guess these checks are duplicated, and this PR removes them. Could you explain where the corresponding code is, please?

File: optuna/samplers/nsgaii/_sampler.py [hunk 3]
```diff
@@ -104,6 +103,14 @@ class NSGAIISampler(BaseSampler):
                 versions without prior notice. See
                 https://github.com/optuna/optuna/releases/tag/v2.5.0.
 
+        elite_population_selection_strategy:
+            A process to select individuals to survive from current population pool. Defaults to
+            :class:`~optuna.samplers.nsgaii._elite_population_selection_strategy.NSGAIIElitePopulationSelectionStrategy`.
```
Comment (toshihikoyanase) [inline optuna/samplers/nsgaii/_sampler.py:108]: The term "process" is also used in the context of `multiprocessing` and `subprocess`, which could potentially lead to confusion. Could you rephrase the explanation, please?

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR. Let me share my early comments.

File: optuna/samplers/_nsgaiii.py (hunk 2 above)
Comment (Alnusjaponica) [reply optuna/samplers/_nsgaiii.py:115]: I should have done this #4806, but these checks were moved to initializer of `NSGAIIChildGenerationStrategy`. 
https://github.com/optuna/optuna/blob/65712ad8e23661305dab902654471be1ccf36fa8/optuna/samplers/nsgaii/_child_generation_strategy.py#L29-L38

File: optuna/samplers/nsgaii/_sampler.py (hunk 3 above)
Comment (Alnusjaponica) [reply optuna/samplers/nsgaii/_sampler.py]: Remove the word "process" from the docstring. Please take another look to confirm if the new one make sense.

File: optuna/samplers/_nsgaiii.py (hunk 2 above)
Comment (toshihikoyanase) [reply optuna/samplers/_nsgaiii.py:115]: I understand. Thanks!

**Review by toshihikoyanase — COMMENTED** [review]
I have a comment on the reference. I guess the strategies are still private.

File: optuna/samplers/nsgaii/_sampler.py (hunk 3 above)
Comment (toshihikoyanase) [reply optuna/samplers/nsgaii/_sampler.py]: Thanks!

File: optuna/samplers/nsgaii/_sampler.py [hunk 4]
```diff
@@ -104,8 +103,17 @@ class NSGAIISampler(BaseSampler):
...
                 https://github.com/optuna/optuna/releases/tag/v2.5.0.
 
+        elite_population_selection_strategy:
+            The selection strategy for determining the individuals to survive from the current
+            population pool. Default to
+            :class:`~optuna.samplers.nsgaii._elite_population_selection_strategy.NSGAIIElitePopulationSelectionStrategy`.
+
+        child_generation_strategy:
+            The strategy for generating child parameters from parent trials. Defaults to
+            :class:`~optuna.samplers.nsgaii._after_trial_strategy.NSGAIIChildGenerationStrategy`.
```
[the remark this answers is not in the document]
Comment (Alnusjaponica) [reply optuna/samplers/nsgaii/_sampler.py]: Thanks for the comment. I updated the docstrings.

File: optuna/samplers/nsgaii/_sampler.py [hunk 5]
```diff
@@ -104,9 +103,23 @@ class NSGAIISampler(BaseSampler):
...
-            A process to be conducted after each trial. Defaults to
-            :class:`~optuna.samplers.nsgaii._after_trial_strategy.NSGAIIAfterTrialStrategy`.
+            A set of procedure to be conducted after each trial. Defaults to :obj:`None`.
+
+            .. note::
+
+                The arguments ``elite_population_selection_strategy``,
+                ``child_generation_strategy``, and ``after_trial_strategy`` was added in v3.3.0 as
+                an experimental feature. The interface may change in newer versions without prior
+                notice. See https://github.com/optuna/optuna/releases/tag/v3.3.0.
```
Comment (toshihikoyanase) [inline optuna/samplers/nsgaii/_sampler.py:122]: It seems the note has been added to only one argument, specifically `after_trial_strategy`, and not to all three. While it might seem redundant, would it be worth considering adding similar notes to `elite_population_selection_strategy` and `child_generation_strategy` as well?

image

**Review by HideakiImamura — APPROVED** [review]
LGTM except @toshihikoyanase's comment.

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. The `readthedocs` CI job failed, but I believe it is unrelated to the changes in this PR. LGTM!
